File: backend/modules/topic_catalog/public.py
"""
Topic Catalog Module - Public Interface

Protocol definition and dependency injection provider.
This is the only interface other modules should import from.
"""

from typing import Protocol

from modules.content.public import ContentProvider

from .service import (
    BrowseTopicsResponse,
    CatalogStatistics,
    RefreshCatalogResponse,
    SearchTopicsResponse,
    TopicCatalogService,
    TopicDetail,
    TopicSummary,
)


class TopicCatalogProvider(Protocol):
    """Protocol defining the topic catalog module's public interface."""

    def browse_topics(self, user_level: str | None = None, limit: int = 100) -> BrowseTopicsResponse: ...
    def get_topic_details(self, topic_id: str) -> TopicDetail | None: ...
    def search_topics(
        self,
        query: str | None = None,
        user_level: str | None = None,
        min_duration: int | None = None,
        max_duration: int | None = None,
        ready_only: bool = False,
        limit: int = 100,
        offset: int = 0,
    ) -> SearchTopicsResponse: ...
    def get_popular_topics(self, limit: int = 10) -> list[TopicSummary]: ...
    def get_catalog_statistics(self) -> CatalogStatistics: ...
    def refresh_catalog(self) -> RefreshCatalogResponse: ...


def topic_catalog_provider(content: ContentProvider) -> TopicCatalogProvider:
    """
    Dependency injection provider for topic catalog services.

    Args:
        content: Content service instance (built with same session as caller).

    Returns:
        TopicCatalogService instance that implements the TopicCatalogProvider protocol.
    """
    return TopicCatalogService(content)


__all__ = [
    "BrowseTopicsResponse",
    "CatalogStatistics",
    "RefreshCatalogResponse",
    "SearchTopicsResponse",
    "TopicCatalogProvider",
    "TopicDetail",
    "TopicSummary",
    "topic_catalog_provider",
]

File: mobile/modules/topic_catalog/public.ts
/**
 * Topic Catalog Public Interface
 *
 * The only interface other modules should import from.
 * Pure forwarder - no logic, just selects/forwards service methods.
 */

import { TopicCatalogService } from './service';
import { TopicCatalogRepo } from './repo';
import type {
  TopicSummary,
  TopicDetail,
  BrowseTopicsResponse,
  TopicFilters,
  CatalogStatistics,
  PaginationInfo,
} from './models';

// Public interface protocol
export interface TopicCatalogProvider {
  // Only expose what learning_session actually needs
  getTopicDetail(topicId: string): Promise<TopicDetail | null>;

  // Internal methods for topic_catalog's own screens (not cross-module)
  browseTopics(
    filters?: TopicFilters,
    pagination?: Omit<PaginationInfo, 'hasMore'>
  ): Promise<BrowseTopicsResponse>;
  searchTopics(
    query: string,
    filters?: TopicFilters,
    pagination?: Omit<PaginationInfo, 'hasMore'>
  ): Promise<BrowseTopicsResponse>;
  getPopularTopics(limit?: number): Promise<TopicSummary[]>;
  getCatalogStatistics(): Promise<CatalogStatistics>;
  refreshCatalog(): Promise<{
    refreshedTopics: number;
    totalTopics: number;
    timestamp: string;
  }>;
  checkHealth(): Promise<boolean>;
}

// Service instance (singleton)
let serviceInstance: TopicCatalogService | null = null;

function getServiceInstance(): TopicCatalogService {
  if (!serviceInstance) {
    const repo = new TopicCatalogRepo();
    serviceInstance = new TopicCatalogService(repo);
  }
  return serviceInstance;
}

// Public provider function
export function topicCatalogProvider(): TopicCatalogProvider {
  const service = getServiceInstance();

  // Pure forwarder - no logic
  return {
    getTopicDetail: service.getTopicDetail.bind(service),
    browseTopics: service.browseTopics.bind(service),
    searchTopics: service.searchTopics.bind(service),
    getPopularTopics: service.getPopularTopics.bind(service),
    getCatalogStatistics: service.getCatalogStatistics.bind(service),
    refreshCatalog: service.refreshCatalog.bind(service),
    checkHealth: service.checkHealth.bind(service),
  };
}

// Export types for other modules
export type {
  TopicSummary,
  TopicDetail,
  BrowseTopicsResponse,
  TopicFilters,
  CatalogStatistics,
  PaginationInfo,
} from './models';
